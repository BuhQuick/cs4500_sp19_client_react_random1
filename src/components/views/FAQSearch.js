import React from 'react'

const FAQSearch = (props) => {
  return (
    <tr>
      <th>
        <div id="searchButtons">
            <button
              id="search"
              disabled={props.disabled}
              onClick={() => props.searchFn()}>
              Search</button>
            <button
              id="clearSearch"
              disabled={!props.disabled}
              onClick={() => props.clearFn()}>
            Clear Search</button>
        </div>
    </th>
  </tr>
  )
}

export default FAQSearch;
