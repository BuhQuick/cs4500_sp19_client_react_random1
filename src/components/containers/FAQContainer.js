import React from 'react'
import FAQService from '../../services/FAQService'
import FAQSearch from '../views/FAQSearch'
import FAQTable from '../views/FAQTable'
import FAQPagination from '../views/FAQPagination'

class FAQContainer extends React.Component {
    constructor(props) {
        super(props)
        this.faqService = FAQService.getInstance()
        this.state = {
            faqs: [],
            searchButtonDisabled: false,
            pageNumber: 1,
            itemsPerPage: 10
        }
        // bind functions here
        this.refreshFAQs = this.refreshFAQs.bind(this)
        this.searchFAQs = this.searchFAQs.bind(this)
        this.clearSearch = this.clearSearch.bind(this)
        this.deleteFAQ = this.deleteFAQ.bind(this)
        this.nextPage = this.nextPage.bind(this)
        this.previousPage = this.previousPage.bind(this)
        this.onChangeItemsPerPage = this.onChangeItemsPerPage.bind(this)
        this.createPageButtons = this.createPageButtons.bind(this)
    }

    componentDidMount() {
        this.faqService
            .findAllFAQs()
            .then(results =>
                this.setState({
                    faqs: results.slice(),
                    searchButtonDisabled: false
                })
            )
    }

    refreshFAQs() {
        this.faqService.findAllFAQs().then(results =>
            this.setState({
                faqs: results
            }))
    }

    deleteFAQ(id) {
        console.log('Delete FAQ Click happened');
        this.faqService.deleteFAQ(id)
            .then(() => this.refreshFAQs());
    }


    searchFAQs() {
        const titleInput = document.getElementById('titleInput');
        const questionInput = document.getElementById('questionInput');
        var title = titleInput.value
        var question = questionInput.value
        debugger;
        if(title === "" && question === "") {
            this.faqService
                .findAllFAQs()
                .then(results =>
                    this.setState(prevState => ({
                        faqs: results,
                        searchButtonDisabled: false
                    }))
                )
            return;
        }
        this.faqService
            .filterFAQs({title: title, question: question})
            .then(results =>
                this.setState(prevState => ({
                    faqs: results,
                    searchButtonDisabled: !prevState.searchButtonDisabled
                }))
            )
    }
    clearSearch() {
        const titleInput = document.getElementById('titleInput');
        const questionInput = document.getElementById('questionInput');
        titleInput.value = ""
        questionInput.value = ""
        this.faqService
            .findAllFAQs()
            .then(results =>
                this.setState(prevState => ({
                    faqs: results,
                    searchButtonDisabled: false
                }))
            )
    }

    onChangeItemsPerPage() {
        var currButtons = document.getElementById("pageButtons").innerHTML = "";
        var eID = document.getElementById("selector");
        var pageText = eID.options[eID.selectedIndex].text;

        if (pageText == "10") {
            this.setState({itemsPerPage: 10, pageNumber: 1})
        }
        if (pageText == "25") {
            this.setState({itemsPerPage: 25, pageNumber: 1})
        }
        if (pageText == "50") {
            this.setState({itemsPerPage: 50, pageNumber: 1})
        }
        if (pageText == "All") {
            this.setState({itemsPerPage: 1000000, pageNumber: 1})
        }
    }
    previousPage() {
        if (this.state.pageNumber != 1) {
            var temp = this.state.pageNumber - 1;
            this.setState({pageNumber: temp })
        }
    }
    nextPage() {
        var totalPages = Math.ceil(this.state.faqs.length / this.state.itemsPerPage);
        if (this.state.pageNumber + 1 <= totalPages) {
            var temp = this.state.pageNumber + 1;
            this.setState({pageNumber: temp })
        }
    }

    createPageButtons() {
        var totalPages = Math.ceil(this.state.faqs.length / this.state.itemsPerPage);
        var i;
        for (i = 1; i < totalPages + 1; i++) {
            if ((document.getElementById("Button " + i)) == null)
            {
                console.log(i);
                var button = document.createElement("BUTTON");
                var buttonText = document.createTextNode("" + i);
                button.appendChild(buttonText);
                button.type = "button";
                button.id = "Button " + i;
                button.value = i;
                button.onclick = function(e) {
                    this.setState({pageNumber: e.target.value})
                }.bind(this);
                document.getElementById("pageButtons").appendChild(button);
            }
        }
    }

    render() {
        let result = (
            <div>
                <FAQTable
                    faqs={this.state.faqs}
                    deleteFn={this.deleteFAQ}
                    pageNumber={this.state.pageNumber}
                    itemsPerPage={this.state.itemsPerPage}/>
                <FAQSearch
                    disabled={this.state.searchButtonDisabled}
                    searchFn={this.searchFAQs}
                    clearFn={this.clearSearch}
                />
                <FAQPagination id="Pagination"
                    previousFn={this.previousPage}
                    nextFn={this.nextPage}
                    changeItemsPerPageFn={this.onChangeItemsPerPage}
                    createButtonsFn={this.createPageButtons}/>
            </div>
        )
        return result;
    }
}

export default FAQContainer;
