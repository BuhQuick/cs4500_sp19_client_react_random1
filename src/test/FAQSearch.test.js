import React from 'react';
import FAQSearch from '../components/views/FAQSearch';
import FAQContainer from '../components/containers/FAQContainer';
import renderer from 'react-test-renderer';

it('Render FAQSearch buttons without Search disabled', () => {
  const tree = renderer.create(
    <FAQSearch
    disabled={false}
    searchFn={FAQContainer.searchFAQs}
    clearFn={FAQContainer.clearSearch}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Render FAQSearch buttons with Search disabled', () => {
  const tree = renderer.create(
    <FAQSearch
    disabled={true}
    searchFn={FAQContainer.searchFAQs}
    clearFn={FAQContainer.clearSearch}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
