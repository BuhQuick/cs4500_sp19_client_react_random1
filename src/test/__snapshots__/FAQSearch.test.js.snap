// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render FAQSearch buttons with Search disabled 1`] = `
<tr>
  <th>
    <div
      id="searchButtons"
    >
      <button
        disabled={false}
        id="search"
        onClick={[Function]}
      >
        Search
      </button>
      <button
        disabled={true}
        id="clearSearch"
        onClick={[Function]}
      >
        Clear Search
      </button>
    </div>
  </th>
</tr>
`;

exports[`Render FAQSearch buttons without Search disabled 1`] = `
<tr>
  <th>
    <div
      id="searchButtons"
    >
      <button
        disabled={false}
        id="search"
        onClick={[Function]}
      >
        Search
      </button>
      <button
        disabled={true}
        id="clearSearch"
        onClick={[Function]}
      >
        Clear Search
      </button>
    </div>
  </th>
</tr>
`;
